### PR TITLE
feat: use `addTimingsCallback` instead of `addPostFrameCallback` to determine app start end

### PR DESCRIPTION
**Pull request by buenaflor**
## :scroll: Description

Improve app start measurements

## :bulb: Motivation and Context

The `addTimingsCallback` contains the accurate data for the first frame unlike `addPostFramecallback` which is not correct as it depends on when `SentryFlutter.init` is called

`addTimingsCallback` is always reported immediately for the first frame so it's very suitable to use here, see https://api.flutter.dev/flutter/scheduler/SchedulerBinding/addTimingsCallback.html

## :green_heart: How did you test it?
Unit tests, manual tests

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [x] No new PII added or SDK only sends newly added PII if `sendDefaultPii` is enabled
- [ ] I updated the docs if needed
- [x] All tests passing
- [x] No breaking changes

### CODE REVIEW

File: flutter/lib/src/integrations/native_app_start_integration.dart [hunk 1]
```diff
@@ -62,8 +72,12 @@ class NativeAppStartIntegration extends Integration<SentryFlutterOptions> {
         if (options.automatedTestMode) {
           rethrow;
         }
+      } finally {
+        _frameCallbackHandler.removeTimingsCallback(timingsCallback);
```
Comment (buenaflor) [inline flutter/lib/src/integrations/native_app_start_integration.dart:76]: we only care about the first frame timing so after we processed it we can just remove the callback

File: flutter/test/fake_frame_callback_handler.dart [hunk 2]
```diff
@@ -1,33 +1,33 @@
...
   }
 
   @override
-  bool hasScheduledFrame = true;
+  void addTimingsCallback(TimingsCallback callback) {
+    timingsCallback = callback;
+  }
 
   @override
-  Future<void> get endOfFrame => Future<void>.value();
```
Comment (buenaflor) [inline flutter/test/fake_frame_callback_handler.dart:32]: removed dead code here

**Comment by buenaflor** [comment]
cc @stefanosiano @denrase please take a look

File: flutter/lib/src/integrations/native_app_start_integration.dart [hunk 3]
```diff
@@ -28,15 +29,24 @@ class NativeAppStartIntegration extends Integration<SentryFlutterOptions> {
...
+      // Set immediately to prevent multiple executions
+      // we only care about the first frame
+      _allowProcessing = false;
+
       try {
         DateTime? appStartEnd;
         if (options.autoAppStart) {
           // ignore: invalid_use_of_internal_member
-          appStartEnd = options.clock();
+          appStartEnd = DateTime.fromMicrosecondsSinceEpoch(timings.first
```
Comment (denrase) [inline flutter/lib/src/integrations/native_app_start_integration.dart:48]: Can we always assume `timings` is not empty? If not, we'd crash here accessing `timings.first`. Maybe we should check if it's not empty, just to be safe.

File: flutter/lib/src/integrations/native_app_start_integration.dart (hunk 1 above)
Comment (denrase) [reply flutter/lib/src/integrations/native_app_start_integration.dart:76]: Can we be sure that `timingsCallback` here is the same reference as the outside `timingsCallback`? I did try and it does work as expected.

Bildschirmfoto 2024-11-18 um 11 10 30

File: flutter/lib/src/integrations/native_app_start_integration.dart (hunk 3 above)
Comment (denrase) [reply flutter/lib/src/integrations/native_app_start_integration.dart:48]: Ok, saw that the tests handle this case. I assume we catch the exception and this is the intended way. 👍

**Review by denrase — APPROVED** [review]
Looking good. Only had 2 comment, but those cases work fine when i tested in the example app. 👍

File: flutter/lib/src/integrations/native_app_start_integration.dart (hunk 1 above)
Comment (buenaflor) [reply flutter/lib/src/integrations/native_app_start_integration.dart:76]: yes but let me add an additional widget test for this

File: flutter/lib/src/integrations/native_app_start_integration.dart (hunk 3 above)
Comment (buenaflor) [reply flutter/lib/src/integrations/native_app_start_integration.dart:48]: yup, we catch it and the app start won't trigger

**Pull request merged**